feat: add `update_metadata_properties` in `FeedbackDataset` and `RemoteFeedbackDataset`

alvarobartt opened the pull request:
# Description

This PR adds the `update_metadata_property` function in the Python SDK to call the `PATCH /api/v1/metadata-properties/{metadata_property_id}` endpoint, also adding the `update` method to the `ArgillaMetadataPropertiesMixin` that contains some common functionality for the remote `metadata_properties` i.e. `metadata_properties` from a `RemoteFeedbackDataset` which is a `FeedbackDataset` in Argilla.

So on, now anyone with either `owner` or `admin` roles can perform updates of the `metadata_properties` via assignment and then push those updates to Argilla via `update_metadata_properties`. Note that the schemas define an in-place method named `_update` which applies to any of `RemoteTermsMetadataProperty`, `RemoteIntegerMetadataProperty`, and `RemoteFloatMetadataProperty`, on any or both of the `title` value and the `visible_for_annotators` flag.

Closes #3960 

## Code-snippet

```python
import argilla as rg

rg.init(api_url="...", api_key="...")

dataset = rg.FeedbackDataset(
    fields=[rg.TextField(name="required-field", required=True), rg.TextField(name="optional-field", required=False)],
    questions=[rg.TextQuestion(name="question", required=True)],
    metadata_properties=[
        rg.TermsMetadataProperty(name="terms-metadata", title="old-title", values=["a", "b", "c"]),
        rg.IntegerMetadataProperty(name="int-metadata", min=0, max=10),
        rg.FloatMetadataProperty(name="float-metadata", min=0.0, max=10.0),
    ],
)

metadata_property = dataset.metadata_property_by_name(name="terms-metadata")
metadata_property.title = "new-title"
metadata_property.visible_for_annotators = False

remote_dataset = dataset.push_to_argilla(name="my-dataset", workspace="my-workspace")
print(remote_dataset.metadata_properties)

remote_metadata_property = remote_dataset.metadata_property_by_name(name="terms-metadata")
remote_metadata_property.title = "old-title"
remote_metadata_property.visible_for_annotators = True

remote_dataset.update_metadata_properties(remote_metadata_property)
print(remote_dataset.metadata_properties)
```

> **⚠️ Warning:** the code below will fail on `RemoteFeedbackDataset` because the `metadata_properties` are fetched every time those are accessed and the values are not persisted in Argilla until one calls `update_metadata_properties` with one or multiple `metadata_properties` with updated attributes.

```python
import argilla as rg

rg.init(api_url="...", api_key="...")

dataset = rg.FeedbackDataset(
    fields=[rg.TextField(name="required-field", required=True), rg.TextField(name="optional-field", required=False)],
    questions=[rg.TextQuestion(name="question", required=True)],
    metadata_properties=[
        rg.TermsMetadataProperty(name="terms-metadata", title="old-title", values=["a", "b", "c"]),
        rg.IntegerMetadataProperty(name="int-metadata", min=0, max=10),
        rg.FloatMetadataProperty(name="float-metadata", min=0.0, max=10.0),
    ],
)

remote_dataset = dataset.push_to_argilla(name="my-dataset", workspace="my-workspace")

remote_dataset.metadata_properties[0].title = "new-title"
assert remote_dataset.metadata_properties[0].title == "new-title"

for remote_metadata_property in remote_dataset.metadata_properties:
   remote_metadata_property.title = "new-title"
assert all(
    remote_metadata_property.title == "new-title" for remote_metadata_property in remote_dataset.metadata_properties
)
```

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [x] Add integration tests for `update_metadata_property` in the Python SDK
- [x] Add unit and integration tests for `update_metadata_properties` in the Python client
- [x] Add integration tests for `Remote{Terms,Integer,Float}MetadataProperty._update` in the Python client

**Checklist**

- [x] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [x] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

alvarobartt committed:
feat: add `update_metadata_property` in SDK

alvarobartt committed:
feat: add `ArgillaMetadataPropertiesMixin.update`

alvarobartt committed:
feat: add `_RemoteMetadataProperty.update`

alvarobartt committed:
test(integration): `update_metadata_property`

alvarobartt committed:
fix: `self` re-assignment on `update`

alvarobartt committed:
test(integration): in-place `update`

alvarobartt committed:
docs: update `CHANGELOG.md`

frascuchon commented on src/argilla/client/feedback/schemas/remote/metadata.py:
```diff
@@ -28,13 +28,48 @@
+            raise ValueError(
+                f"Provided `visible_for_annotators={visible_for_annotators}` is the same as the current one"
+                f" `{self.visible_for_annotators}`, you need either to provide a different one, or just pass it empty."
+            )
```
I think we should just skip the remote update when the original and new data are the same. We should avoid side effects as much as possible. For example, running a property update twice, the first will do an update but the second will raise an error, which can be confusing to the user.

Just skipping the remote update when data remains the same is a good optimization, but not a reason to raise an error IMHO

alvarobartt replied:
Hmm ok then I will change that to `warnings.warn` instead, and skip the call to update if nothing needs to be updated

frascuchon commented on tests/integration/client/sdk/v1/test_metadata_properties.py:
```diff
@@ -53,3 +53,30 @@ async def test_delete_metadata_property(self, role: UserRole, db: "AsyncSession"
         await db.refresh(dataset, attribute_names=["metadata_properties"])
         assert len(dataset.metadata_properties) == 0
+
+    @pytest.mark.parametrize("role", [UserRole.owner, UserRole.admin])
+    async def test_update_metadata_property(self, role: UserRole, db: "AsyncSession") -> None:
+        dataset = await DatasetFactory.create()
+        await TermsMetadataPropertyFactory.create(dataset=dataset)
+        await FloatMetadataPropertyFactory.create(dataset=dataset)
+        await IntegerMetadataPropertyFactory.create(dataset=dataset)
+        user = await UserFactory.create(role=role, workspaces=[dataset.workspace])
+
+        api = Argilla(api_key=user.api_key, workspace=dataset.workspace.name)
+
+        await db.refresh(dataset, attribute_names=["metadata_properties"])
+        assert len(dataset.metadata_properties) == 3
+
+        for metadata_property in dataset.metadata_properties:
+            response = update_metadata_property(
+                client=api.client.httpx, id=metadata_property.id, title="new-title", visible_for_annotators=False
+            )
+            assert response.status_code == 200
+            assert response.parsed.title == "new-title"
+            assert response.parsed.visible_for_annotators is False
+
+        await db.refresh(dataset, attribute_names=["metadata_properties"])
+        assert len(dataset.metadata_properties) == 3
+        for metadata_property in dataset.metadata_properties:
+            assert metadata_property.title == "new-title"
+            assert metadata_property.visible_for_annotators is False
```
Avoid using test factories for integration tests. Also, I think this test is mostly covered in the `test_medatata.py` module

If you want to check API call details, you can create a unit test with a mocked HTTP client to check the request send

alvarobartt replied:
But note that's a Python SDK test, not a Python client test, shouldn't we keep the factories then?

alvarobartt committed:
fix: `update` warns when no updates

alvarobartt committed:
test(integration): split `pytest.warns` from `pytest.raises`

alvarobartt committed:
refactor(test): `TestSuiteMetadataPropertiesSDK` to unit

alvarobartt committed:
fix(style): clear imports and remove unused

alvarobartt committed:
revert: `update` requires pre-assignment

As proposed by @frascuchon, in order to be aligned with the current approach for the rest of the updates, those should be done via assignment instead of via `update` args

Co-authored-by: Paco Aranda <francis@argilla.io>

frascuchon commented on src/argilla/client/feedback/schemas/remote/metadata.py:
```diff
@@ -28,13 +29,32 @@
+        validate_assignment = True
+
+    @allowed_for_roles(roles=[UserRole.owner, UserRole.admin])
+    def update(self: "AllowedRemoteMetadataPropertyTypes") -> None:
```
Do we need a type definition for `self`? Normally it would be an instance of `_RemoteMetadataProperty` which can be enough, right?

frascuchon replied:
It is a bit weird for me define specialized classes ("AllowedRemoteMetadataPropertyTypes" is a union of specialized classes of `_RemoteMetadataProperty`) at the base class level

alvarobartt replied:
As it's just internally used it's there for the type-hints, that won't affect neither the users nor the user experience, so I think we can leave it there, as that's beneficial for us internally and has zero cost

frascuchon commented on src/argilla/client/feedback/schemas/remote/metadata.py:
```diff
@@ -28,13 +29,32 @@
+
     @allowed_for_roles(roles=[UserRole.owner, UserRole.admin])
-    def delete(self) -> "AllowedMetadataPropertyTypes":
+    def delete(self: "AllowedRemoteMetadataPropertyTypes") -> "AllowedMetadataPropertyTypes":
```
ditto

frascuchon commented on tests/unit/client/sdk/v1/metadata_properties/test_api.py:
```diff
@@ -0,0 +1,50 @@
+
+@pytest.mark.asyncio
+class TestSuiteMetadataPropertiesSDK:
+    async def test_update_metadata_property(self, mock_httpx_client: httpx.Client) -> None:
```
I think we dont need asyncio support for these test suite
```suggestion
class TestSuiteMetadataPropertiesSDK:
    def test_update_metadata_property(self, mock_httpx_client: httpx.Client) -> None:
```

alvarobartt replied:
True, I'll undo that!

alvarobartt committed:
test(style): `pytest.mark.asyncio` not needed

alvarobartt committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' of github.com:argilla-io/argilla into feat/metadata-property-update

alvarobartt committed:
hotfix: add missing `__init__.py`

alvarobartt committed:
feat: `allow_mutation=False` for `id` and `client`

alvarobartt committed:
refactor: `_RemoteMetadataProperty` methods protected

alvarobartt committed:
feat: define `abstractmethod`s for `metadata_properties`

alvarobartt committed:
fix(chore): add missing `__init__.py`

alvarobartt committed:
feat: `update_metadata_properties` for `{Remote}FeedbackDataset`

alvarobartt committed:
test(integration): `update_metadata_properties`

alvarobartt committed:
refactor(test): move `test_local` to `local/test_dataset`

alvarobartt committed:
fix(test): add `update_metadata_properties` in `TestFeedbackDataset`

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

alvarobartt committed:
docs: update `CHANGELOG.md`

frascuchon commented on src/argilla/client/feedback/schemas/remote/metadata.py:
```diff
@@ -28,13 +29,28 @@
+        self.__dict__.update(updated_metadata_property.__dict__)
+
+    @allowed_for_roles(roles=[UserRole.owner, UserRole.admin])
+    def _delete(self: "AllowedRemoteMetadataPropertyTypes") -> "AllowedMetadataPropertyTypes":
```
Since `_update` is not being used. We can remove it. Also, we can remove the `_delete` and use directly the mixin helper for calling the API.

frascuchon commented on tests/integration/client/feedback/dataset/remote/test_dataset.py:
```diff
@@ -366,6 +338,70 @@ async def test_add_metadata_property_sequential(
+
+        for metadata_property in remote_dataset.metadata_properties:
+            assert metadata_property.title == "new-title"
+            assert metadata_property.visible_for_annotators is False
```
These integration tests should be written using public methods. We can rename them to check a bulk update or a one-by-one update.

frascuchon commented on tests/integration/client/feedback/dataset/remote/test_dataset.py:
```diff
@@ -366,6 +338,70 @@ async def test_add_metadata_property_sequential(
+        assert len(remote_dataset.metadata_properties) == len(feedback_dataset.metadata_properties)
+
+        for idx, metadata_property in enumerate(remote_dataset.metadata_properties, 1):
+            metadata_property._delete()
```
same here

frascuchon commented on tests/integration/client/feedback/schemas/remote/test_metadata.py:
```diff
@@ -77,7 +77,71 @@ def test_create_and_restore(
+            ),
+        ],
+    )
+    def test__update(
```
If we're not using the `_update` method, this test can be removed

frascuchon commented on tests/integration/client/feedback/schemas/remote/test_metadata.py:
```diff
@@ -77,7 +77,71 @@ def test_create_and_restore(
+            assert metadata_property.title == title
+            assert metadata_property.visible_for_annotators == visible_for_annotators
+
+    def test__delete(self, owner: "UserModel") -> None:
```
same here if we remove the `_delete` method

frascuchon committed:
chore: Remove unused methods and unused base class

frascuchon committed:
chore: Using mixin for metadata deletion

frascuchon committed:
chore: Remove tests using removed methods

frascuchon committed:
chore: Renaming integration tests and avoid using private methods

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci
